spring-cloud-gcp-data-firestore/src/main/java/org/springframework/cloud/gcp/data/firestore/util/ObservableReactiveUtil.java

```diff
@@ -82,19 +83,26 @@ private ObservableReactiveUtil() {
 	 * @param <T> type of input objects that will be wrapped in {@link RequestT}
 	 * @return A {@link Flux} of server responses
 	 */
-	public static <ResponseT, RequestT, T> Flux<ResponseT> streamingBidirectionalCall(
+	public static <ResponseT, RequestT, T> Flux<T> streamingBidirectionalCall(
 			Function<StreamObserver<ResponseT>, StreamObserver<RequestT>> initialCall,
-			Flux<T> inputs, BiFunction<T, Flux<ResponseT>, Mono<RequestT>> responseHandler) {
+			Flux<List<T>> inputs,
+			BiFunction<List<T>, ResponseT, RequestT> responseHandler) {
+		AtomicReference<StreamObserver<RequestT>> requestStream = new AtomicReference<>();
 
-		AtomicReference<StreamObserver<RequestT>> requestObserver = new AtomicReference<>();
-		Flux<ResponseT> responses = (Flux<ResponseT>) streamingCall(
-				obs -> requestObserver.set(initialCall.apply((StreamObserver<ResponseT>) obs))).cache();
+		Flux<ResponseT> responsesFlux =
+				ObservableReactiveUtil
+						.<ResponseT>streamingCall(obs -> requestStream.set(initialCall.apply(obs)))
+						.cache(1);
 
 		return inputs
-				.flatMap((T input) -> responseHandler.apply(input, responses)
-						.doOnNext(request -> requestObserver.get().onNext(request)))
-				.doOnComplete(() -> requestObserver.get().onCompleted())
-				.thenMany(responses);
+				.flatMap(input ->
+						responsesFlux.next()
+								.map(mostRecentResponse -> responseHandler.apply(input, mostRecentResponse))
+								.doOnNext(nextRequest -> requestStream.get().onNext(nextRequest))
+								.thenReturn(input))
+				.doOnComplete(() -> requestStream.get().onCompleted())
+				.delayUntil(input -> responsesFlux)
```

Comment:
I don't think you want to use delayUntil here because it will repeatedly check if responsesFlux has completed (delayUntil checks this condition for every single element). 

You could just use thenMany here since it'll just wait until your inputs are all sent over and then it'll wait until responseFlux is done, and then it'll continue on to your flatMap on your next line.

Reply:
Right, thenMany would be ideal but then I lose my original `input` type.

I think ideally we'd have something like a `delayUntilMany`... or something. But I think delayUntil is still the best choice here.

Reply:
you're completely right